### ISSUE TITLE
Swarm mode reports invalid hostname if hostname is changed after Docker daemon is started

### ISSUE THREAD

**Issue by nathanleclaire** [docker/docker#27173]
**Description**

If the hostname of a machine is changed before Docker is started and after swarm mode is enabled, Swarm mode seems to report the old (invalid) hostname.

**Steps to reproduce the issue:**

You dupe this fairly simply with `docker-machine` and two node cluster (probably even with one really).

First `create`.  Machine sets host name according to machine name automatically.  And Docker is started on these machines of course.

Then change host name on created machines.

Then start swarm init / join.

Here's a little script you can use to dupe if you want.  You should see that `hostname` command will report the correct set hostname when executed after the middle steps to set it FWIW.

``` console
docker-machine create -d virtualbox manager
docker-machine create -d virtualbox worker
docker-machine ssh worker 'sudo hostname baz && sudo sed -i "s/worker/baz/g" /etc/hosts && echo baz | sudo tee /etc/hostname'
docker-machine ssh manager 'sudo hostname quux && sudo sed -i "s/manager/quux/g" /etc/hosts && echo quux | sudo tee /etc/hostname'
docker-machine ssh manager docker swarm init --advertise-addr eth1
docker-machine ssh worker docker swarm join --token $(docker-machine ssh manager docker swarm join-token -q worker) $(docker-machine ip manager):2377
docker-machine ssh manager docker node ls
```

Cool, the hostname was set before swarm mode was initiated, so the new hostnames should be propagated when the swarm is started, right?

`docker node ls` reports the old values though.  These should say `quux` and `baz` for hostname.

```
$ docker-machine ssh manager docker node ls
ID                           HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS
3of2on83v5orfuozfqpnq00ul *  manager   Ready   Active        Leader
5usitwuan5g1n7l8drx15kue9    worker    Ready   Active
```

**Describe the results you received:**

Hostname is not reported correctly.

**Describe the results you expected:**

Hostname is reported correctly.

**Additional info**

If Docker daemon is restarted _before_ the Swarm initialization steps it seems to report the hostnames correctly.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64
```

(I seem to have duped this with `1.12.2-rc2` as well though)

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay null bridge host
Swarm: active
 NodeID: 3of2on83v5orfuozfqpnq00ul
 Is Manager: true
 ClusterID: eqisp01jd77ahunfx80asdwez
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.99.105
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.17-boot2docker
Operating System: Boot2Docker 1.12.1 (TCL 7.2); HEAD : ef7d0b4 - Thu Aug 18 21:18:06 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.9 MiB
Name: quux
ID: M7Q6:DORB:VZZB:WCJO:LVIG:247Y:BQCV:FZGH:VRMN:Y77H:XKA5:PECS
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 32
 Goroutines: 124
 System Time: 2016-10-05T23:37:30.070842443Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
Insecure Registries:
 127.0.0.0/8
```

**Cute animal photo**

Ladies and gentlemen, I present to you, the Hostname Corgi.

![image](https://cloud.githubusercontent.com/assets/1476820/19135573/6f8db3b2-8b1a-11e6-9b7f-b43164c6b7eb.png)

Thanks.

cc @aluzzardi @stevvooe @aaronlehmann FYI

**Comment by aaronlehmann** [comment]
ping @nishanttotla - does the dynamic node updates PR fix this?

**Comment by nishanttotla** [comment]
@aaronlehmann it should be supported. In the SwarmKit agent, we explicitly override hostname in the node description if it exists, so we should catch any updates.
```go
// nodeDescriptionWithHostname retrieves node description, and overrides hostname if available
func (a *Agent) nodeDescriptionWithHostname(ctx context.Context) (*api.NodeDescription, error) {
	desc, err := a.config.Executor.Describe(ctx)

	// Override hostname
	if a.config.Hostname != "" && desc != nil {
		desc.Hostname = a.config.Hostname
	}
	return desc, err
}
```
I'll try to reproduce this issue.

**Comment by nishanttotla** [comment]
@nathanleclaire I managed to reproduce this on 1.12.2 on Ubuntu 16.04, so this seems like a legitimate issue. `docker info` reports the updated hostname correctly, but not `docker node ls` or `docker node inspect`. It does seem like hostname is not being picked up by SwarmKit, I'll look into it.

**Comment by nishanttotla** [comment]
I think I found the root cause. Getting errors in the logs that indicate that the node update isn't working, because an updated description is never found:
```bash
ERRO[0020] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0040] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0060] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0080] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0100] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0120] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0140] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
```
The error originates here
```go
desc, err := a.config.Executor.Describe(ctx)
```
The error indicates an HTTP response to HTTPS client. Figuring out why this is happening.

**Comment by aaronlehmann** [comment]
@nishanttotla: Are you using go 1.6? There was a problem in the vendored version of the Docker API client that only shows up with versions of go below 1.7. It has since been fixed in Docker master.

I assume you're running swarmd/swarmctl? I don't think this error happens with Swarm mode in Docker since it doesn't use the Docker API to communicate with the daemon.

**Comment by nishanttotla** [comment]
@aaronlehmann I am using go 1.6, but trying with 1.7 now to see if the issue goes away. Unfortunately, the issue does seem to happen with Swarm mode in Docker, with the following version:
```bash
Client:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:29:41 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:29:41 2016
 OS/Arch:      linux/amd64
```
I'm making sure that the hostname is set correctly through SwarmKit.

**Comment by nishanttotla** [comment]
Also, something is really strange:
```bash
root@testlet:~/work/src/github.com/docker/swarmkit# hostname
newtestletFOUR
root@testlet:~/work/src/github.com/docker/swarmkit# docker swarm init --advertise-addr 45.55.23.195
Swarm initialized: current node (1st1mpfq8ugpewycgucdt29of) is now a manager.

To add a worker to this swarm, run the following command:

    docker swarm join \
    --token SWMTKN-1-4nppjc4acktb9gh2qh4k1trlu9v2r1ut1kgjwb8ayooycwjdlx-b78qjsfx9hnnxfgwv0b8fgzd5 \
    45.55.23.195:2377

To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.

root@testlet:~/work/src/github.com/docker/swarmkit# docker node ls
ID                           HOSTNAME       STATUS  AVAILABILITY  MANAGER STATUS
1st1mpfq8ugpewycgucdt29of *  newtestletONE  Ready   Active        Leader
```
`newtestletONE` was an old hostname, the current one is `newtestletFOUR`

**Comment by nishanttotla** [comment]
I confirmed that this is an issue in Docker. It seems like the updated node description isn't actually reaching the agent (I suspect that dynamic node updates are probably not working at all in Docker, although they work perfectly fine with SwarmKit).

While `SystemInfo` actually picks up the updated hostname, the agent still sees the old one. I'll have to investigate more, but I'd say this is important to fix before 1.13.

cc @aaronlehmann @stevvooe @aluzzardi @nathanleclaire

**Comment by nishanttotla** [comment]
I figured out the issue, finally.

Will send out a SwarmKit PR with a proposed fix.
